### ISSUE TITLE
take advantage of certifi.io to locate the CA cert file

### ISSUE THREAD

**Issue by bear** [errbotio/errbot#763]
Because the default value for XMPP_CA_CERT_FILE is `/etc/ssl/certs/ca-certificate.crt` which is linux flavoured errbot doesn't work properly on any system that isn't linux like OS X.

This could be solved by using certifi.io and it's `pip` installable package to have a clean cert file in place that is not OS dependent.

I did this locally by having in my config.py 

```
import certifi
XMPP_CA_CERT_FILE = certifi.where()
```
but i'm wondering if this should just be used in the core?  If so I will work up a pull request.

thanks

**Comment by gbin** [comment]
It does make sense, please make a PR, we already had a share of support to do on this one :)

**Comment by bear** [comment]
Is this ok to add as a core requirement? (i.e. having certifi listed as a required package in setup.py)

**Comment by zoni** [comment]
I don't think this is a good idea and am quite opposed to this. From a security perspective this just makes an already tricky situation (OS trust stores differ per OS and the subject matter itself, while simple on the surface, requires some knowledge) worse because ___you now have to manage multiple trust stores___. You have your OS trust store, nowadays you have browsers coming bundled with their own, then you have errbot using it's own, etc etc.

This is problematic because, should a global CA become compromised the whole point is to no longer trust that CA is quickly as possible. If there was only one trust store (the one of your OS) this would be easy. You either invalidate the root certificate of the offending CA yourself and you know you're safe, or you wait for your vendor (Microsoft, Apple, your Linux distribution maintainers, etc) to release patches, which should normally come quickly.

It becomes problematic with apps using their own stores. Now you have to know which apps ignore the OS store and then you have to update those as well. As more apps move to this model of using their own, the bigger the chances are that you'll miss one or more of them.

Additionally, if this was automatically installed and used, few people would even realize we are now ignoring their OS's trust store and using our own. They won't know that ***they must*** upgrade certifi in order to receive a new, updated CA bundle and so we expose users to unnecessary risk.

Summing up, I am strongly :-1: on using certify or any other bundled CA certificates.

That said though, only two backends (XMPP -and by extension HipChat, IF using HipChat server- and IRC) even really benefit from having this configurable. The others (Slack, HipChat, Telegram) are cloud services where you never even need to configure a different certificate, because there is no self-hosted version using a custom non-system-trusted CA in the first place.

I think the real issue here is that XMPP and possibly IRC (I'm not sure on the situation there) don't use the system trust store out of the box by default on non-Linux systems. We should fix that core issue instead.

**Comment by gbin** [comment]
@zoni thanks for the articulated response, it makes a lot of sense !
Maybe we could fix it as a simple better default with a platform detection on the default value ?

**Comment by bear** [comment]
That is a great response indeed!  To be honest part of why i'm suggesting that certifi be used is that OS level trust stores are rarely updated and is part of the reason why certifi.io was created.

Sure if your running errbot on your own infra and, like I do as part of my $dayjob as operations, you manage trust stores like other resources it would be as you describe.

I'm more interested in helping the other 70? 50? 20? percent of the folks who, say, are running on Debian or OS X where the CA updates can take months if not years to appear.

I do take your point about fixing the XMPP trust store issue, and to be blunt, since I'm one of the maintainers of SleekXMPP you would think that I could handle that ;)

**Comment by zoni** [comment]
> To be honest part of why i'm suggesting that certifi be used is that OS level trust stores are rarely updated and is part of the reason why certifi.io was created.
> [..]
> I'm more interested in helping the other 70? 50? 20? percent of the folks who, say, are running on Debian or OS X where the CA updates can take months if not years to appear.

Right, this is a very good point. I thought this situation had improved a good deal in recent year(s), are you saying it's still generally problematic today? (I haven't looked into it recently)

I still worry about switching over to certifi though. Even if we do, this started out with setting it for `XMPP_CA_CERT_FILE` which is just one of the various backends we have. Most our other backends build on top of other libraries which may not even offer the ability to change this, instead always using those of the OS.

I fear that unless we can support it on _all_ backends, it would be highly confusing to use a mix of the two.
